### PR TITLE
Prioritize edgelist representations in `to_networkx_graph`

### PR DESCRIPTION
**Pull request by rossbar**
This PR was originally motivated by #7401. The main issue there is the extraneous `ImportWarning` s that are raised by `to_networkx_graph` (often via the graph constructors) when creating graphs in an environment without the default dependencies (numpy, scipy, pandas). This is the case because converting from these data structures is actually attempted *before* attempting to interpret the input as an edgelist. However, creating a graph from an edge list (or edgelist-like, such as an EdgeView or generator of edges) is a very common operation - certainly much more so than from numpy/scipy/pandas objects within NetworkX itself. This latter point can be demonstrated relatively simply: see https://github.com/networkx/networkx/pull/7402#issuecomment-2048157399.

I wanted to investigate whether it'd be possible to reorganize the control-flow of `to_networkx_graph` so that the most common (at least according to the internal NX usage heuristic) data structures are handled first. This has the advantage of solving #7401 for basically all cases[^1]. I suspect it may result in a minor performance improvement too, though the use-case where that's actually meaningful may not be that common (i.e. an analysis where many graph instances are being created). I've added a benchmark to demonstrate this last point as well; results on my machine:

<details>
  <summary> ~10% improvement for this branch </summary>
<pre>
| Change   | Before [d7b61bd1] <main>         | After [a2b6d7fe] <reorder-to_networkx_graph>                              | Ratio   | Benchmark (Parameter)                                                                                                                                  |
|----------|----------------------------|------------------------------------------------|---------|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| -        | 5.54±0.09μs                | 4.95±0.07μs                                    | 0.89    | benchmark_to_networkx_graph.ToNetworkXGraphBenchmark.time_to_networkx_graph_direct(<class 'networkx.classes.graph.Graph'>)                             |
| -        | 3.12±0.01ms                | 2.79±0.02ms                                    | 0.89    | benchmark_to_networkx_graph.ToNetworkXGraphBenchmark.time_to_networkx_graph_direct_multi_instance(<class 'networkx.classes.digraph.DiGraph'>)          |
| -        | 2.74±0.02ms                | 2.46±0.01ms                                    | 0.90    | benchmark_to_networkx_graph.ToNetworkXGraphBenchmark.time_to_networkx_graph_direct_multi_instance(<class 'networkx.classes.graph.Graph'>)              |
| -        | 7.28±0.04μs                | 6.44±0.04μs                                    | 0.88    | benchmark_to_networkx_graph.ToNetworkXGraphBenchmark.time_to_networkx_graph_via_constructor(<class 'networkx.classes.digraph.DiGraph'>)                |
| -        | 6.34±0.03μs                | 5.70±0.03μs                                    | 0.90    | benchmark_to_networkx_graph.ToNetworkXGraphBenchmark.time_to_networkx_graph_via_constructor(<class 'networkx.classes.graph.Graph'>)                    |
|          | 3.54±0.07ms                | 3.24±0.01ms                                    | 0.91    | benchmark_to_networkx_graph.ToNetworkXGraphBenchmark.time_to_networkx_graph_via_constructor_multi_instance(<class 'networkx.classes.digraph.DiGraph'>) |
| -        | 3.18±0.02ms                | 2.85±0.05ms                                    | 0.90    | benchmark_to_networkx_graph.ToNetworkXGraphBenchmark.time_to_networkx_graph_via_constructor_multi_instance(<class 'networkx.classes.graph.Graph'>)     |
</details>

The potential downside of this approach is that there may be subtle corner cases due to the change in the order/"breadth" of the isinstance checks. I'm confident that this works for every case that the NetworkX test suite covers (using the procedure described [here](https://github.com/networkx/networkx/pull/7402#issuecomment-2048157399)), but there may be cases that are uncovered. For example, it turns out there is *a single* test in the test suite that uses a tuple of edges instead of a list of edges. That case would have been missed were it not for that single test!

[^1]: The only exceptions being branches where an exception is explicitly raised.

### CODE REVIEW

File: networkx/convert.py [hunk 1]
```diff
@@ -90,13 +90,6 @@ def to_networkx_graph(data, create_using=None, multigraph_input=False):
         except Exception as err:
             raise nx.NetworkXError("Input is not a correct NetworkX graph.") from err
 
-    # pygraphviz  agraph
```
Comment (rossbar) [inline networkx/convert.py:93]: This is an unfortunately rendering of the diff... FWIW the check has not been modified at all, it was just moved so that the added edgelist check is first!

**Review by jarrodmillman — APPROVED** [review]

**Review by dschult — APPROVED** [review]
Thanks @rossbar !!
Looks good!
